go/consensus/tendermint/full/full.go

```diff
@@ -785,6 +757,37 @@ func (t *fullService) lazyInit() error { // nolint: gocyclo
 			}
 		}()
 
+		// Configure state sync if enabled.
+		var stateProvider tmstatesync.StateProvider
+		if viper.GetBool(CfgConsensusStateSyncEnabled) {
+			t.Logger.Info("state sync enabled")
+
+			if t.p2p == nil {
+				t.Logger.Info("failed to create state sync client", "err", "p2p disabled")
+				return fmt.Errorf("failed to create state sync client: p2p disabled")
+			}
+
+			// Enable state sync in the configuration.
+			tenderConfig.StateSync.Enable = true
+			tenderConfig.StateSync.TrustHash = viper.GetString(CfgConsensusStateSyncTrustHash)
+
+			// Create new state sync state provider.
+			cfg := lightAPI.ClientConfig{
+				GenesisDocument: tmGenDoc,
+				TrustOptions: tmlight.TrustOptions{
+					Period: viper.GetDuration(CfgConsensusStateSyncTrustPeriod),
+					Height: int64(viper.GetUint64(CfgConsensusStateSyncTrustHeight)),
+					Hash:   tenderConfig.StateSync.TrustHashBytes(),
+				},
+			}
+			if stateProvider, t.lightClient, err = newStateProvider(t.ctx, t.genesis.ChainContext(), cfg, t.p2p); err != nil {
```

Comment:
Should probably be moved out, but I didn't yet start modifying the configuration settings. The trust settings configurations will be refactored to something like: `consensus.tendermint.light_client.trust_hash`. And the light client can then be instantiated separately and passed into the state sync if state sync is enabled (via the existing "consensus.tendermint.state_sync.enabled" flag).

Reply:
Hm, that's true.

So if state sync is configured, the light-client would need to be initialized before the consensus node, with trust settings from the configuration, and if state sync is disabled (meaning this is a full node), the light-client should be initialized once the full node is synced, from the trusted root obtained from the node.
Does this make sense?

Probably doesn't make sense for state-sync to use a separate internal light-client? Although that would simplify the above to: always wait for consensus to be synced, then initialize a light client with a trusted latest block. Which would then also nicely lead to decoupling the light-client out of the consensus backend.

Reply:
> meaning this is a full node

In both cases this can be a full node, one just doesn't have all the history (e.g. starts from the state sync snapshot) and the other syncs from genesis.

> the light-client should be initialized once the full node is synced, from the trusted root obtained from the node

Yeah it should probably use the lowest and highest blocks after full node consensus sync known by the node and store them as trusted.

Reply:
> Yeah it should probably use the lowest and highest blocks after full node consensus sync known by the node and store them as trusted.

Would it make sense that the light-client also follows full node blocks and periodically stores the latest block as trusted? (e.g. every 100 blocks)?

Reply:
Hm maybe once per epoch, given that the validator set will not change inside an epoch.

Reply:
Added once per epoch.